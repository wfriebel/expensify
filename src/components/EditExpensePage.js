import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    expenseFormOnSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    removeOnClick = (e) => {
        this.props.startRemoveExpense(this.props.expense.id)
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
            <ExpenseForm 
                expense={this.props.expense}
                onSubmit={this.expenseFormOnSubmit}
            />
            <button onClick={this.removeOnClick} 
            >Remove</button>
        </div>
        );
    };
};

const mapStateToProps = (state, props) => {
   return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
   }
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);