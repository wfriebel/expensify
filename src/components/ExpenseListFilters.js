import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";


export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        const selection = e.target.value;
        if(selection === 'date') {
            this.props.sortByDate();
        } else if (selection === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
       return (
           <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            defaultValue={this.props.filters.text}
                            onChange={this.onTextChange}
                            placeholder="Search Expenses"
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            defaultValue={this.props.filters.sortBy}
                            onChange={this.onSortChange} 
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            startDateId="123"
                            endDateId="123"
                        />
                     </div>
                </div>
            </div>
        );
    }
}

const MapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
})

export default connect(MapStateToProps, mapDispatchToProps)(ExpenseListFilters);