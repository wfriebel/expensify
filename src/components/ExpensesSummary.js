import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expensesTotal';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const amount = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{amount}</span>
                 </h1>
                 <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                 </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: expenses.length,
        expensesTotal: expensesTotal(expenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);