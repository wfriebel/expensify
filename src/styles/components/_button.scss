.button {
    background-color: $blue;
    border: none;
    font-weight: 300;
    padding: $s-size;
    color: white;
    font-size: $font-size-large;
    display: inline-block;
    text-decoration: none;
    line-height: 1;
}

.button--link {
    background-color: transparent;
}

.button--secondary {
    background-color: #888888;
}