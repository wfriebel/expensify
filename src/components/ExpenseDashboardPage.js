import React from 'react';
import ExpenseList from './ExpenseList.js'
import ExpenseListFilters from './ExpenseListFilters.js';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;