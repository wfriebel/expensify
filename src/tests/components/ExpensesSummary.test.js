import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with a count of 1', () => {
    const expensesCount = 1;
    const expensesTotal = 200;
    const wrapper = shallow(<ExpensesSummary
        expensesCount={expensesCount} 
        expensesTotal={expensesTotal}
    />);
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpensesSummary with a count of 0', () => {
    const expensesCount = 0;
    const expensesTotal = 200;
    const wrapper = shallow(<ExpensesSummary
        expensesCount={expensesCount} 
        expensesTotal={expensesTotal}
    />);
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpensesSummary with a count of 2', () => {
    const expensesCount = 2;
    const expensesTotal = 200;
    const wrapper = shallow(<ExpensesSummary
        expensesCount={expensesCount} 
        expensesTotal={expensesTotal}
    />);
    expect(wrapper).toMatchSnapshot();
})