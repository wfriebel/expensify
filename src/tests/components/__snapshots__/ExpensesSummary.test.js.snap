// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with a count of 0 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $2.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with a count of 1 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with a count of 2 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $2.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
